### ISSUE TITLE
Add HighNodeUtilization strategy

### ISSUE THREAD

**Issue by vvbogdanov87** [kubernetes-sigs/descheduler#196]
I want to keep my cluster as compact as possible. In order to do that, I switched kubernetes-scheduler policy from LeastRequestedPriority to MostRequestedPriority. It would be greate to have corresponding HighNodeUtilization strategy for descheduler.

**Comment by NGTOne** [comment]
👍 for this.

**Comment by hanumanthan** [comment]
hi @seanmalloy   `HighNodeUtilization` - Is it the inverse of `LowNodeUtilization` ?
If a node's utilization <= threshold, evict pods from that node such that `MostRequestedPriority` will reschedule it accordingly ?

**Comment by seanmalloy** [comment]
> hi @seanmalloy   `HighNodeUtilization` - Is it the inverse of `LowNodeUtilization` ?
> If a node's utilization <= threshold, evict pods from that node such that `MostRequestedPriority` will reschedule it accordingly ?
> 

@hanumanthan yes that is correct.

**Comment by hanumanthan** [comment]
@seanmalloy I think https://github.com/kubernetes-sigs/descheduler#lownodeutilization should be called as `HighNodeUtilisation` as its evicting pods from nodes which are highly utilised and this feature is `LowNodeUtilisation` - evict pods from nodes that are under utilised 😄 

Interested to pick this up. The strategy would be like
```yaml
strategies:
  "HighNodeUtilization":
     enabled: true
     params:
       nodeResourceUtilizationThresholds:
         thresholds:
           "cpu" : 20
           "memory": 20
           "pods": 20
```
Calculate total utilised resources in a node and if its less than the threshold, consider pods in those nodes for eviction. 
Do we need a similar TargetThreshold ?

**Comment by seanmalloy** [comment]
@hanumanthan thanks for offering to work on this feature.

I believe we should not use the word "utilization" to describe this descheduler strategy. We should instead name it something like `HighNodeAllocation`. I proposed renaming the `LowNodeUtilization` strategy to `LowNodeAllocation` when creating the v1alpha2 API. See this https://github.com/kubernetes-sigs/descheduler/pull/314#issuecomment-638975891.

Maybe something like this:
```
strategies:
  "HighNodeAllocation":
     enabled: true
     params:
       nodeResourceUtilizationThresholds:
         thresholds:
           "cpu" : 20
           "memory": 20
           "pods": 20
```

@damemi and @ingvagabund what do you think? Can we have @hanumanthan start implementing this new strategy? Are there additional details we need to work out before starting to implement this?

**Comment by hanumanthan** [comment]
Great @seanmalloy 
Could you please explain on `threshold` and 'targetThreshold` params - 
Understood the usage  from the code and doc and would like to understand  the reasoning part.  If pods resources on nodes > targetThreshold are considered, what is the significance of `threshold` ?

**Comment by ingvagabund** [comment]
> Are there additional details we need to work out before starting to implement this?

Implementation of both strategies is going to be quite similar. Though, picking different set of nodes to process. Worth refactoring the current `LowNodeUtilization` to allow both modes to avoid code duplication. I'd start that how much the current code can be "reused".

**Comment by ingvagabund** [comment]
`threshold`: any node with utilization under `threshold` is considered as an under-utilized node
`targetThreshold`: any node with utilization above `targetThreshold` is considered as an over-utilized node

**Comment by damemi** [comment]
> `threshold`: any node with utilization under `threshold` is considered as an under-utilized node
> `targetThreshold`: any node with utilization above `targetThreshold` is considered as an over-utilized node

@hanumanthan this explanation is correct (I'm not sure if this is documented).

>  If pods resources on nodes > targetThreshold are considered, what is the significance of threshold?

`LowNodeUtilization` doesn't just look at pods on over-utilized nodes and evict them, it specifically looks for pods on over-utilized nodes that could, in theory, fit onto nodes that are under-utilized (defined by threshold). These terms could be a lot clearer (like `highThreshold` and `lowThreshold` or something)

**Comment by hanumanthan** [comment]
How should the specs be for `HighNodeUtilization`?

Say `lowThreshold=30%` and `highThreshold=80%`, Should it be evict pods in nodes below 30% utilization
1.  place them in nodes between 30% and 80% utilization or
2. place them in nodes above 80% utilization. Evict and place until these nodes reach 100% for max utilization.

I am inclined towards 1 because
1. It keeps more node towards the utilization target
2. Pods dont consume all the allocated resources of the node as even a single pod going slightly above the request resource will get killed

**Comment by ingvagabund** [comment]
Over utilized nodes are more likely prone to memory/cpu pressure. So the first option is not much practical. Which, as you say, leaves only the second option. Also given you don't know where all the evicted pods lends in the end, you can repeat the same process of computing how much of each resource is still left.

**Comment by damemi** [comment]
> How should the specs be for `HighNodeUtilization`?
> 
> Say `lowThreshold=30%` and `highThreshold=80%`, Should it be evict pods in nodes below 30% utilization
> 
> 1. place them in nodes between 30% and 80% utilization or
> 2. place them in nodes above 80% utilization. Evict and place until these nodes reach 100% for max utilization.

The 2nd option is more similar to how `LowNodeUtilization` currently works: it doesn't just evict pods from high utilization nodes to put them on any other node, it specifically evicts pods that it hopes will go onto low utilization nodes. For that reason, I would prefer option 2.


> I am inclined towards 1 because
> 
> 1. It keeps more node towards the utilization target

Your utilization target in this strategy is to pack the pods as tightly as possible onto the over-utilized nodes. If you just want to evict pods from under-utilized nodes, why even have an upper threshold in this case?

> 2. Pods dont consume all the allocated resources of the node as even a single pod going slightly above the request resource will get killed

This would be up to the scheduler to make sure a pod doesn't get scheduled onto a node that it cannot fit on, so it's not a concern

The descheduler is capable of considering all of the pods and deciding which ones can more efficiently fill the nodes up to 100%. If you're running this with `MostRequestedPriority`, that's what you want. If you don't target to specifically land on those nodes, then it's possible you *will* end up evicting pods that can't tightly fit onto the high-usage nodes and your bin-packing strategy suffers.

**Comment by ingvagabund** [comment]
> Your utilization target in this strategy is to pack the pods as tightly as possible onto the over-utilized nodes. If you just want to evict pods from under-utilized nodes, why even have an upper threshold in this case?

The upper bound in here might seem redundant. Though, in case there's e.g. 10 under utilized nodes but only 1 over utilized, you want to evict only as many resources so all the evicted pods can still lend on all nodes while still keeping the nodes from becoming over utilized.

Yes, it's up to the scheduler to score the nodes to concentrate the pods on as few nodes as possible. Yet, when you do it too much, you might trigger pod eviction due to a resource pressure. I don't remember what's the fail safe for making sure none of the nodes get too much saturated. If anyone wants to exercise the fail safe, ignoring the upper threshold or setting it to 100% will do. Though, just in case someone says 95% is fine, have the descheduler evict only as many pods so all the nodes are at most 95% utilized, the upper bound is a must. It does not guarantee the scheduler will respect those 95%, though it might.

**Comment by damemi** [comment]
You're right, I'm not entirely sure that it's necessary for us to set an upper buffer like that but I understand the reasoning and it sounds fine to me.

What I was really wondering was if it's possible to use that upper threshold to make more efficient calculations when choosing pods to evict (like what `LowNodeUtilization` does), but I think there isn't an equivalent because it's just semantically not the same going from low->high as high->low.

**Comment by damemi** [comment]
Bumping, is there anyone working on this right now?

**Comment by sftim** [comment]
Aside: Microsoft Research did a good [technical paper](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/provisioning-nsdi08.pdf) on the broader load skewing concepts; their motivation was to schedule for lowest energy consumption.

**Comment by zorro786** [comment]
Much interested in this feature. I see a pending PR for this anyone working actively on it? @seanmalloy

**Comment by seanmalloy** [comment]
> Much interested in this feature. I see a pending PR for this; anyone working actively on it? @seanmalloy

It looks like PR #341 is not being actively worked on. This issues is up for grabs if someone wants to work on it.

**Comment by hanumanthan** [comment]
@seanmalloy @damemi  Sorry for not communicating my absence in this issue and PR - Had to take a long break due to personal reasons.
I am happy to resume this if no one has picked it up yet. In the PR, its mentioned that #384 has good amount of refractor. Shall I take that as base and open a new PR or #341 is still good to continue ?

**Comment by seanmalloy** [comment]
> #341

@hanumanthan I don't think anyone is actively working on it, so feel free to continue your work. Thanks!

**Comment by hanumanthan** [comment]
Validating my understanding ( as I have been out of touch for a long while)
Which approach to select

##### Approach 1
We need two thresholds
1. lower threshold - to identify nodes that have usage `<= threshold`
2. upper threshold - to identify nodes that have usage `> lower threshold` and `<= upper threshold`

evict pods from low usage nodes only if they can fit in nodes that have usage below upper threshold. In this case, scheduler with `MostRequestedPriority` will pack the pods on nodes as much as possible concentrating it to smallest number of nodes

##### Approach 2
1. lower threshold - to identify nodes that have usage `<= threshold`
2. upper threshold - identify nodes that have usage `> threshold`

evict pods from low usage nodes such that they can fit in high usage nodes. If the high threshold is set as 100%, then there is nothing that descheduler can evict.

**Comment by damemi** [comment]
Thanks for picking this back up @hanumanthan 

I still think the upper threshold could lead to some issues. From what @ingvagabund said above answering my questions about it:
> you want to evict only as many resources so all the evicted pods can still lend on all nodes while still keeping the nodes from becoming over utilized.

the problem I see is that the scheduler will try to compact the pods onto nodes as tightly as possible, without any concept of "overutilized" in the context we are using it. So, we risk making different decisions than the scheduler. Since the descheduler is designed entirely with optimistic assumptions of the scheduler, any potential divergence between the 2 should be done carefully.

I think the risk of this is low, since we are making educated evictions by calculating what can actually fit where, but it could be possible with the right(wrong) threshold configurations that the performance isn't as expected.

I'm okay continuing with that approach if others think it's the way to go, but we should thoroughly test it and clearly document the intent of this strategy 🙂

**Comment by ingvagabund** [comment]
We may add the upper threshold later once there's a request for it. In case someone gets to a scenario where the strategy is evicting too many pods and causing memory/cpu pressure too often. It's possible the scheduler will be able to deal with this situation actually and balance the load across multiple high utilized nodes.

I'd say let's pin the upper threshold to 100%, so the problem of

> lower threshold - to identify nodes that have usage <= lower threshold
> upper threshold - to identify nodes that have usage > lower threshold and <= upper threshold
>
> evict pods from low usage nodes only if they can fit in nodes that have usage below upper threshold.

gets translated from into

- **nodes to evict pods from**: to identify nodes that have usage <= lower threshold
- **nodes to evict pods to (prediction)**: to identify nodes that have usage > lower threshold

evict pods from low usage nodes only if they can fit in remaining nodes that have usage above lower threshold.

The upper bound is here to create a "reserve" so the highly utilized nodes do not get overwhelmed too much. Later on (if requested) we can solve the following case:

- **nodes to evict pods from**: to identify nodes that have usage <= lower threshold
- **nodes to evict pods to (prediction)**: to identify nodes that have usage > lower threshold

evict pods from low usage nodes only if they can fit in remaining nodes that have usage above lower threshold **but bellow higher threshold** (as the original **Approach 1**).

**Comment by damemi** [comment]
This scenario is what I am worried about when configuring an upper threshold for this strategy:
* Lower threshold is 25%, upper is 75%
* **NodeA** is **underutilized** with only 1 pod using 10% of its memory
* **Victim pod** is evicted with Descheduler assuming it can only fit on **Node B** since **Node C** is already "**overutilized**"
* Scheduler actually sees **Node C** as the better candidate, since that will be the more effective bin packing node, and the pod can fit on it

![highutilization](https://user-images.githubusercontent.com/1839101/113881702-aab1eb80-978a-11eb-9b91-f784ada108ce.jpg)

So if we provide a configurable upper threshold, with the intent of saying "pods will only be scheduled onto nodes that won't go over this threshold" I am not sure if we can even guarantee that behavior.

We obviously don't want to try to evict onto nodes that would go >100%, so I think this will still need an internal upper theshold pinned to 100% (but that is an implementation detail).

**Comment by ingvagabund** [comment]
@hanumanthan do you have all the information you need to proceed?

**Comment by hanumanthan** [comment]
Yes, I will raise a PR in a couple of days

**Comment by caiohasouza** [comment]
Hi @hanumanthan, great job on PR #550! It was exactly what I was needing, do you have some release estimate?

**Comment by damemi** [comment]
Hi @caiohasouza, we're still reviewing that PR but it should hopefully get into the v0.22 Descheduler release at the latest (we are currently waiting to release v0.21)

**Comment by caiohasouza** [comment]
Hi @damemi cool, hmm, so this may take a few months? just so I can get an idea of time...

**Comment by damemi** [comment]
@caiohasouza Unfortunately yes, we try to follow the main Kubernetes release cycle, which we are currently past-due on for 1.21 while we wait for the Kind 1.21 release (more info in our release tracking issue https://github.com/kubernetes-sigs/descheduler/issues/491)

If we can get #550 merged before then, it will be available in the 1.21 release (likely only a few more weeks). Otherwise, it won't be published until the next release, which is scheduled for sometime around July.

You do have the option of building your own images from the master branch in the mean time and running those, but that branch isn't strictly supported and there is a chance of bugs.

**Comment by caiohasouza** [comment]
@damemi i got it, I will wait for the official release and hope it comes out in version 1.21m thank you!!

**Comment by just1900** [comment]
Hi, I'm wondering what's the difference between this "HighNodeUtilization" strategy and "Scale-down" feature provided by (cluster-autoscaler)[https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-does-scale-down-work]?

Seems like both are triggered by resource request, and both will evict the pods.

**Comment by damemi** [comment]
@just1900 I think I answered your question on slack, but to post it here too for others:

I think you're right on the main difference being that descheduler doesn't actually scale anything down. This can lead to some different use cases, for example descheduler's strategy here is more targeted at bin packing while CA tries to also bin pack, but with as few nodes as possible.

It looks like another difference is that descheduler has 2 configurable parameters, for a high threshold and a low threshold. This creates a middle ground of appropriate utilization, so you can set a target range rather than trying to pack everything as tightly as possible.

Hope that helps
